ext/nnstreamer/android_source/gstamcsrc.c

```diff
@@ -0,0 +1,987 @@
+          priv->isPlaying = FALSE;
+
+          Looper_post (priv->looper, MSG_CODEC_PAUSE_ACK, self, TRUE);
+
+          LOGI ("Paused");
+        }
+        break;
+      case MSG_CODEC_PAUSE_ACK:
+        /** Only used to flush buffers */
+        break;
+      case MSG_CODEC_RESUME:
+        if (!priv->isPlaying) {
+          priv->renderstart = -1;
+          priv->isPlaying = TRUE;
+
+          Looper_post (priv->looper, MSG_CODEC_BUFFER, self, FALSE);
+
+          LOGI ("Resumed");
+        }
+        break;
+      default:
+        LOGE ("Unexpected message: cmd(%d)", cmd);
+        break;
+    }
+  }
+}
+
+/**
+ * @brief start function, called when state changed null to ready
+ * start the media codec and its looper
+ */
+static gboolean
+gst_amc_src_start (GstBaseSrc * src)
+{
+  GstAMCSrc *self = GST_AMC_SRC (src);
+  GstAMCSrcPrivate *priv = GST_AMC_SRC_GET_PRIVATE (self);
+
+  /** setup a looper */
+  priv->looper = Looper_new();
+  Looper_set_handle (priv->looper, looper_handle);
+
+  priv->previous_ts = GST_CLOCK_TIME_NONE;
+  priv->started = TRUE;
+
+  /** start media codec */
+  AMediaCodec_start(priv->codec);
+  Looper_post (priv->looper, MSG_CODEC_BUFFER, self, false);
+
+  return TRUE;
+}
+
+/**
+ * @brief stop function, called when state changed ready to null
+ */
+static gboolean gst_amc_src_stop (GstBaseSrc * src)
+{
+  GstAMCSrc *self = GST_AMC_SRC (src);
+  GstAMCSrcPrivate *priv = GST_AMC_SRC_GET_PRIVATE (self);
+
+  g_mutex_lock (&priv->mutex);
+
+  gst_data_queue_flush (priv->outbound_queue);
+  priv->started = FALSE;
+
+  g_mutex_unlock (&priv->mutex);
+
+  return TRUE;
+}
+
+/**
+ * @brief check if source supports seeking
+ */
+gboolean gst_amc_src_is_seekable (GstBaseSrc * src)
+{
+  GstAMCSrc *self = GST_AMC_SRC_CAST (src);
+  GstAMCSrcPrivate *priv = GST_AMC_SRC_GET_PRIVATE (self);
+
+  return priv->seekable;
+}
+
+/**
+ * @brief callback for checking data_queue full
+ */
+static gboolean
+data_queue_check_full_cb (GstDataQueue * queue, guint visible,
+    guint bytes, guint64 time, gpointer checkdata)
+{
+  /** it's dummy */
+  return FALSE;
+}
+
+/**
+ * @brief initialize amcsrc element.
+ */
+static void
+gst_amc_src_init (GstAMCSrc * self)
+{
+  GstAMCSrcPrivate *priv = GST_AMC_SRC_GET_PRIVATE (self);
+
+  priv->outbound_queue = gst_data_queue_new (data_queue_check_full_cb, NULL, NULL, NULL);
+  priv->filename = NULL;
+
+  priv->width = 0;
+  priv->height = 0;
+  priv->framerate = 0; 
+  priv->duration = 0;
+
+  g_mutex_init (&priv->mutex);
+
+  gst_base_src_set_format (GST_BASE_SRC (self), GST_FORMAT_TIME);
+  gst_base_src_set_live (GST_BASE_SRC (self), TRUE);
+  gst_base_src_set_async (GST_BASE_SRC (self), TRUE);
+}
+
+/**
+ * @brief finalize the instance
+ */
+static void
+gst_amc_src_finalize (GObject * object)
+{
+  GstAMCSrc *self = GST_AMC_SRC (object);
+  GstAMCSrcPrivate *priv = GST_AMC_SRC_GET_PRIVATE (self);
+
+  Looper_post (priv->looper, MSG_CODEC_DONE, self, TRUE);
```

Comment:
Would it be better to call `AMediaCodec_stop()` in `_stop()`
and delete the media codec object with `AMediaCodec_delete()` in `_finalize()`?

Reply:
`_start()` and `_stop()` are called when going upward and downward the states.
As `_start()` creates the looper and starts media codec, I think `looper_exit()` should be done in `_stop()`

Reply:
I see. I think the logic of termination should be changed. thanks!